src/blockchain_db/lmdb/db_lmdb.cpp

```diff
@@ -351,98 +283,177 @@ typedef struct outtx {
     uint64_t local_index;
 } outtx;
 
+const std::string& mdb_databases::name(source database)
+{
+  static std::array<std::string, 18> names{
+    "blocks",
+    "block_heights",
+    "block_info",
+
+    "output_txs",
+    "output_amounts",
+
+    "txs",
+    "txs_pruned",
+    "txs_prunable",
+    "txs_prunable_hash",
+    "txs_prunable_tip",
+    "tx_indices",
+    "tx_outputs",
+
+    "spent_keys",
+
+    "txpool_meta",
+    "txpool_blob",
+
+    "alt_blocks",
+
+    "hf_versions",
+
+    "properties",
+  };
+
+  return names.at(static_cast<uint8_t>(database));
+}
+
+MDB_stat mdb_databases::stat(MDB_txn* txn, source database) const
+{
+  MDB_stat stat;
+  auto res = mdb_stat(txn, get(database), &stat);
+
+  if (res == 0)
+    return stat;
+
+  auto error_message =
+    "Failed to query " +
+    mdb_databases::name(database) +
+    ": " +
+    mdb_strerror(res);
+
+  throw0(DB_ERROR(error_message.c_str()));
+}
+
+void mdb_databases::drop(MDB_txn* txn, source database, bool close)
+{
+  auto res = mdb_drop(txn, get(database), close);
+
+  if (res == 0)
+    return;
+
+  auto error_message =
+    "Failed to drop " +
+    mdb_databases::name(database) +
+    ": " +
+    mdb_strerror(res);
+
+  throw0(DB_ERROR(error_message.c_str()));
+}
+
 std::atomic<uint64_t> mdb_txn_safe::num_active_txns{0};
 std::atomic_flag mdb_txn_safe::creation_gate = ATOMIC_FLAG_INIT;
 
-mdb_threadinfo::~mdb_threadinfo()
+void lmdb_resized(MDB_env *env)
 {
-  MDB_cursor **cur = &m_ti_rcursors.m_txc_blocks;
-  unsigned i;
-  for (i=0; i<sizeof(mdb_txn_cursors)/sizeof(MDB_cursor *); i++)
-    if (cur[i])
-      mdb_cursor_close(cur[i]);
-  if (m_ti_rtxn)
-    mdb_txn_abort(m_ti_rtxn);
+  mdb_txn_safe::prevent_new_txns();
```

Comment:
Since your cleaning this up already ... several lines can throw (including each log statement), making this call exception unsafe. There's `epee::on_scoped_exit` (which can be re-written to not allocate) that would do that, or a couple of other usual techniques.

Reply:
The simplest (and safest) way is probably to use an `std::mutex` and `std::unique_lock` (or their boost equivalents). That way the destructor of `std::unique_lock` will ensure the lock is released.

Ignoring the possible exceptions thrown from logging, however, I wonder what the correct course of action would be here. If the environment is resized and `mdb_env_set_mapsize` fails then we cannot use lmdb at that point (because another process may be writing outside of our allotted map size). This would mean that even if we correctly release the lock here, we'd be left in an unusable state (the next call will probably fail again with MDB_MAP_RESIZED).

Does it make sense to retry a failing `mdb_env_set_mapsize`?

Reply:
The scheme is actually a reader_writer_lock, but it might take some refactoring to get there. This is also (basically) guaranteed to be a user-space spinlock, but everyone has differing opinions on those.